crates/noirc_evaluator/src/ssa/anchor.rs

```diff
@@ -0,0 +1,263 @@
+use std::collections::{HashMap, VecDeque};
+
+use super::{
+    context::SsaContext,
+    mem::ArrayId,
+    node::{NodeId, ObjectType, Opcode, Operation},
+};
+
+#[derive(Clone, Debug)]
+pub enum MemItem {
+    NonConst(NodeId),
+    Const(usize),     //represent a bunch of memory instructions having constant index
+    ConstLoad(usize), //represent a bunch of load instructions having constant index
+}
+
+#[derive(Debug)]
+pub enum CseAction {
+    ReplaceWith(NodeId),
+    Remove(NodeId),
+    Keep,
+}
+
+/// A list of instructions with the same Operation variant
+#[derive(Default, Clone)]
+pub struct Anchor {
+    map: HashMap<Opcode, HashMap<Operation, NodeId>>, //standard anchor
+    cast_map: HashMap<NodeId, HashMap<crate::node::ObjectType, NodeId>>, //cast anchor
+    mem_map: HashMap<ArrayId, Vec<VecDeque<(usize, NodeId)>>>, //Memory anchor: one Vec for each array where Vec[i] contains the list of load and store instructions having index i, and the mem_item position in which they appear
+    mem_list: HashMap<ArrayId, VecDeque<MemItem>>, // list of the memory instructions, per array, and grouped into MemItems
+}
+
+impl Anchor {
+    pub fn push_cast_front(&mut self, operation: &Operation, id: NodeId, res_type: ObjectType) {
+        match operation {
+            Operation::Cast(cast_id) => {
+                let mut by_type = HashMap::new();
+                by_type.insert(res_type, id);
+                self.cast_map.insert(*cast_id, by_type);
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn push_front(&mut self, operation: &Operation, id: NodeId) {
+        let op = operation.opcode();
+
+        match operation {
+            Operation::Cast(_) => unreachable!(),
+            _ => {
+                if let std::collections::hash_map::Entry::Vacant(e) = self.map.entry(op) {
+                    let mut prev_list = HashMap::new();
+                    prev_list.insert(operation.clone(), id);
+                    e.insert(prev_list);
+                } else {
+                    self.map.get_mut(&op).unwrap().insert(operation.clone(), id);
+                }
+            }
+        }
+    }
+
+    pub fn find_similar_instruction(&self, operation: &Operation) -> Option<NodeId> {
+        let op = operation.opcode();
+        if self.map.contains_key(&op) && self.map[&op].contains_key(operation) {
+            return Some(self.map[&op][operation]);
+        }
+        None
```

Comment:
`self.map.get(&op).and_then(|inner| inner.get(operation).copied())` a bit awkward, but saves the repeated lookups

Reply:
Done